socket.go

```diff
@@ -555,6 +555,62 @@ func (socket *mongoSocket) Query(ops ...interface{}) (err error) {
 	return err
 }
 
+func (socket *mongoSocket) QueryRaw(payload []byte) (data []byte, err error) {
+	// Buffer is ready for the pipe.  Lock, allocate ids, and enqueue.
+	socket.Lock()
+	if socket.dead != nil {
+		dead := socket.dead
+		socket.Unlock()
+		debugf("(QueryRaw)Socket %p to %s: failing query, already closed: %s", socket, socket.addr, socket.dead.Error())
+		return nil, dead
+	}
+
+	requestId := socket.nextRequestId + 1
+	if requestId == 0 {
+		requestId++
+	}
+
+	socket.nextRequestId = requestId + 1
+	// request ID always start at 4th byte in payload according to
+	// mongodb wired protocol
+	// Although we could just use request IDs from original payload
+	// but better to just use own request IDs
+	setInt32(payload, 4, int32(requestId))
+
+	var wait, change sync.Mutex
+	var replyDone bool
+	var replyData []byte
+	var replyErr error
+	wait.Lock()
+	socket.replyFuncs[requestId] = func(err error, reply *replyOp, docNum int, docData []byte) {
+		change.Lock()
+		if !replyDone {
+			replyDone = true
+			replyErr = err
+			if err == nil {
+				replyData = docData
+			}
+		}
+		change.Unlock()
+		wait.Unlock()
```

Comment:
the nesting of this `wait.Unlock()` call seems a little bit strange to me.  I'm not really familiar enough with Go to understand what this anonymous function is doing.  But having the `wait.Unlock()` nested inside an anonymous function when the `wait.Lock()` is called from a different scope seems dubious to me.

Reply:
I would most likely not write code this way if I have to achieve this, but I copied code how other queries are done by mgo which is well tested code path (See SimpleQuery code). There are definitely better and simpler ways to write this code.

replyFunc is executed as a callback when data is read back from socket after writing payload to socket. `wait` and `change` locks serve purpose of making sure that data is read from socket before copying data for return from buffer. `wait` mutex is captured in anonymous function. Anonymous function unlocks after it's done copying data from callback parameter (socket response). Actual query function will wait until this exchange has happened. change.Lock and unlock methods in query function are redundant here, but kept to keep code same as in SimpleQuery (where those are needed). If it is not clear, it is much easier to explain such kind of locking in person :) I'll add some comments in code though.